kotlin/perform-valve-isolation-trace/src/main/java/com/esri/arcgisruntime/sample/performvalveisolationtrace/MainActivity.kt

```diff
@@ -0,0 +1,296 @@
+    fab.setOnClickListener {
+      fab.isExpanded = !fab.isExpanded
+    }
+
+    // close the options sheet when the map is tapped
+    mapView.onTouchListener = object : DefaultMapViewOnTouchListener(this, mapView) {
+      override fun onTouch(view: View?, event: MotionEvent?): Boolean {
+        if (fab.isExpanded) {
+          fab.isExpanded = false
+        }
+        return super.onTouch(view, event)
+      }
+    }
+
+    // create and load the utility network
+    map.addDoneLoadingListener {
+      createUtilityNetwork()
+    }
+  }
+
+  /**
+   * Creates a utility network from the string resource url and loads it.
+   * Initializes the starting point and UI.
+   */
+  private fun createUtilityNetwork(
+  ) {
+    // create a graphics overlay for the starting location and add it to the map view
+    val startingLocationGraphicsOverlay = GraphicsOverlay().also {
+      mapView.graphicsOverlays.add(it)
+    }
+
+    // create and apply renderers for the starting point graphics overlay
+    val startingPointSymbol = SimpleMarkerSymbol(
+      SimpleMarkerSymbol.Style.CROSS,
+      Color.GREEN,
+      25f
+    )
+    startingLocationGraphicsOverlay.renderer = SimpleRenderer(startingPointSymbol)
+
+    // create a utility network from the url and load it
+    val utilityNetwork = UtilityNetwork(getString(R.string.utility_network_url), mapView.map)
+    utilityNetwork.loadAsync()
+    utilityNetwork.addDoneLoadingListener {
+      if (utilityNetwork.loadStatus == LoadStatus.LOADED) {
+        // get a trace configuration from a tier
+        val networkDefinition = utilityNetwork.definition
+        val domainNetwork = networkDefinition.getDomainNetwork("Pipeline")
+        val tier = domainNetwork.getTier("Pipe Distribution System")
+        val traceConfiguration = tier.traceConfiguration
+
+        // create a trace filter
+        traceConfiguration.filter = UtilityTraceFilter()
+
+        // get a default starting location
+        val networkSource = networkDefinition.getNetworkSource("Gas Device")
+        val assetGroup = networkSource.getAssetGroup("Meter")
+        val assetType = assetGroup.getAssetType("Customer")
+        val startingLocation = utilityNetwork.createElement(
+          assetType,
+          UUID.fromString("98A06E95-70BE-43E7-91B7-E34C9D3CB9FF")
+        )
+
+        // get the first feature for the starting location, and get its geometry
+        val elementFeaturesFuture =
+          utilityNetwork.fetchFeaturesForElementsAsync(listOf(startingLocation))
+
+        elementFeaturesFuture.addDoneListener {
+          try {
+            val startingLocationFeatures = elementFeaturesFuture.get()
+
+            if (startingLocationFeatures.isNotEmpty()) {
+              val startingLocationGeometry = startingLocationFeatures[0].geometry
+
+              if (startingLocationGeometry is Point) {
+                val startingLocationGeometryPoint = startingLocationFeatures[0].geometry as Point
```

Comment:
These lines can be expressed more succinctly with
```suggestion
// get the first feature for the starting location as a point
(startingLocationFeatures[0] as? Point)?.let {startingLocationGeometryPoint ->
```

Reply:
ooh, that's clean